include/dlaf/eigensolver/gen_to_std/impl.h

```diff
@@ -89,11 +90,70 @@ void gemmPanelUpdateTile(Executor&& ex, hpx::shared_future<matrix::Tile<const T,
+  hpx::dataflow(executor_hp, matrix::unwrapExtendTiles(tile::hegst_o), 1, blas::Uplo::Upper,
+                std::move(a_kk), std::move(u_kk));
+}
+
+template <class Executor, Device device, class T>
+void trsmPanelTile(Executor&& executor_hp, hpx::shared_future<matrix::Tile<const T, device>> u_kk,
+                   hpx::future<matrix::Tile<T, device>> a_ki) {
+  hpx::dataflow(executor_hp, matrix::unwrapExtendTiles(tile::trsm_o), blas::Side::Left,
+                blas::Uplo::Upper, blas::Op::ConjTrans, blas::Diag::NonUnit, T(1.0), u_kk,
+                std::move(a_ki));
+}
+
+template <class Executor, Device device, class T>
+void hemmPanelTile(Executor&& executor_hp, hpx::shared_future<matrix::Tile<const T, device>> a_kk,
+                   hpx::shared_future<matrix::Tile<const T, device>> u_ki,
+                   hpx::future<matrix::Tile<T, device>> a_ki) {
+  hpx::dataflow(executor_hp, matrix::unwrapExtendTiles(tile::hemm_o), blas::Side::Left,
+                blas::Uplo::Upper, T(-0.5), a_kk, u_ki, T(1.0), std::move(a_ki));
+}
+
+template <class Executor, Device device, class T>
+void her2kTrailingDiagTile(Executor&& ex, hpx::shared_future<matrix::Tile<const T, device>> a_ki,
+                           hpx::shared_future<matrix::Tile<const T, device>> u_ki,
+                           hpx::future<matrix::Tile<T, device>> a_ii) {
+  hpx::dataflow(ex, matrix::unwrapExtendTiles(tile::her2k_o), blas::Uplo::Upper, blas::Op::ConjTrans,
+                T(-1.0), a_ki, u_ki, BaseType<T>(1.0), std::move(a_ii));
+}
+
+template <class Executor, Device device, class T>
+void gemmTrailingMatrixTile(Executor&& ex, hpx::shared_future<matrix::Tile<const T, device>> mat_ki,
+                            hpx::shared_future<matrix::Tile<const T, device>> mat_kj,
+                            hpx::future<matrix::Tile<T, device>> a_ij) {
+  hpx::dataflow(ex, matrix::unwrapExtendTiles(tile::gemm_o), blas::Op::ConjTrans, blas::Op::NoTrans,
+                T(-1.0), mat_ki, mat_kj, T(1.0), std::move(a_ij));
+}
+
+template <class Executor, Device device, class T>
+void trsmPanelUpdateTile(Executor&& executor_hp, hpx::shared_future<matrix::Tile<const T, device>> u_ii,
+                         hpx::future<matrix::Tile<T, device>> a_ki) {
+  hpx::dataflow(executor_hp, matrix::unwrapExtendTiles(tile::trsm_o), blas::Side::Right,
+                blas::Uplo::Upper, blas::Op::NoTrans, blas::Diag::NonUnit, T(1.0), u_ii,
+                std::move(a_ki));
+}
+
+template <class Executor, Device device, class T>
+void gemmPanelUpdateTile(Executor&& ex, hpx::shared_future<matrix::Tile<const T, device>> a_ki,
+                         hpx::shared_future<matrix::Tile<const T, device>> u_ij,
+                         hpx::future<matrix::Tile<T, device>> a_kj) {
+  hpx::dataflow(ex, matrix::unwrapExtendTiles(tile::gemm_o), blas::Op::NoTrans, blas::Op::NoTrans,
+                T(-1.0), a_ki, u_ij, T(1.0), std::move(a_kj));
+}
+}
```

Comment:
IMHO the tradeoff between code duplication and algorithm mapping here has dropped dramatically.

By giving a quick look (I hope I'm not missing something relevant), what changes between the functions in namespaces `gentostd_l` and `gentostd_u` are:
- naming of arguments (`l` for `u`, algorithm mapping for indices)
- `blas::Uplo::Upper` for `blas::Uplo::Lower` for BLAS/LAPACK calls
- some other paramerers of the BLAS/LAPACK calls depends on previous point (e.g. NoTrans/ConjTrans, Left/Right)

IMHO It would be ok to drop the first bullet point of the previous list of changes, i.e. can be a bit more "generic" to suite both variants, and the second point can be addressed differently, e.g. by passing a template parameter and putting the logic for depending parameters inside the function (and the call will become `trsmPanelUpdateTile<Uplo::Lower>(...)`).

To me both solutions does not look perfect, so I look forward to others feedbacks.

@rasolca @msimberg  @teonnik what do you think?

Reply:
I think this approach is the more readable.
This case is still quite simple, but imagine if you have to combine uplo and op cases:
do you prefer:
```
LN:
function(....)
  blas(hardcoded parameters, ...)
LT:
...
UN:
...
UT:
...
```
or
```
template <uplo, op>
function(...) //(or similarly without template function(uplo, op, ...))
  blas_param1 = (uplo == Lower) && (op==NoTrans) || (uplo == Upper) && (op!=NoTrans) ? value1 : value2
  ...
  blas(blas_param1, ...).
```?

Reply:
Note: If I would apply your argument (code duplication) for algorithms as well, we should have a single implementation in which indices/panels/communications/computations depends on the value of the parameters.

Reply:
My note was not generic, but very specific for this case. I agree that in other cases, like the one of your example, the tradeoff changes in favor of having a custom version for each call.

> Note: If I would apply your argument (code duplication) for algorithms as well, we should have a single implementation in which indices/panels/communications/computations depends on the value of the parameters.

This is an extreme view of the thing, I didn't mean that.

However, I'm ok with this solution, as said in my first comment. I was just wondering about possible solutions.

Discussion